### PR TITLE
Subqueries

### LINKED ISSUE #33: Write tutorial for Using subqueries
Task
Write tutorial for Using subqueries

Starter file

[Using subqueries](https://ploomber-sql.readthedocs.io/en/latest/advanced-querying-techniques/subqueries.html)

Use existing scripts to download and parse data, try to use the db with multiple tables

### PR DESCRIPTION
**Pull request by jpjon**
## Describe your changes
Created the first draft for the subqueries section of the course.
## Issue number
#33 
Closes #33 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)

----
:books: Documentation preview :books:: https://ploomber-sql--36.org.readthedocs.build/en/36/

### CODE REVIEW

**Comment by jpjon** [comment]
First draft ready for review.

File: course-material/advanced-querying-techniques/subqueries.md [hunk 1]
```diff
@@ -1,14 +1,316 @@
...
+```{code-cell} ipython3
+%%sql
+SELECT *
+FROM s1.loan
+LIMIT 5
+```
+
+### Subquery in SELECT
+
+The query belows results in the average loan amount for each "account_id" by having a subquery in the `SELECT` clause.
```
Comment (TegveerG) [inline course-material/advanced-querying-techniques/subqueries.md:114]: In my notebooks, I have been highlighting variables with `variable`, I don't mind using quotes too. A minor catch but I think it's important for consistency across the course. @lfunderburk what do you think?

File: course-material/advanced-querying-techniques/subqueries.md [hunk 2]
```diff
@@ -1,14 +1,316 @@
...
+      (SELECT AVG(l.amount) 
+      FROM s1.loan AS l 
+      WHERE l.account_id = a.account_id
+      ) AS average_loan_amount
+FROM s1.account AS a
+WHERE average_loan_amount IS NOT NULL
+ORDER BY average_loan_amount
+```
+
+Let's first focus on the "outer query" of this statement. The outer query is everything in the SQL statement besides the second argument in the `SELECT` clause. This outer query results in just the account_id from the `s1.account` table. The "inner query" is the second argument in the `SELECT` statement. This query is called for each row from the outer query, which finds the average loan amount for each account by relation in the inner query's `WHERE` clause.
```
Comment (TegveerG) [inline course-material/advanced-querying-techniques/subqueries.md:143]: account_id -> "account_id" or `account_id`, depending on Laura

File: course-material/advanced-querying-techniques/subqueries.md [hunk 3]
```diff
@@ -1,14 +1,316 @@
...
+  WHERE status = 'A'
+)
+```
+
+With this query, we are able to see all loans that are greater than the average loan amount for 'A' status loans.
+
+## You try: Use JupySQL to perform the queries and answer the questions.
+
+### Question 1 (Medium):
+Find all account_id's that have a loan alongside with their loan status, their loan amount, and the average loan amount for their loan's loan status. Order by account_id.
```
Comment (TegveerG) [inline course-material/advanced-querying-techniques/subqueries.md:190]: Same idea as the previous comments here for variable names

File: course-material/advanced-querying-techniques/subqueries.md [hunk 4]
```diff
@@ -1,14 +1,316 @@
...
+
+
+<!-- #region -->
+<details>
+
+<summary>Answers</summary>
+
+The difficult part of this question lies in the second `INNER JOIN`. This second `INNER JOIN` uses a subquery to have the average amount of each 'status'. This information is used when calling an the last `INNER JOIN` on the outer query's 'status' variable.
```
Comment (TegveerG) [inline course-material/advanced-querying-techniques/subqueries.md:198]: typo probably->"an"

File: course-material/advanced-querying-techniques/subqueries.md [hunk 5]
```diff
@@ -1,14 +1,316 @@
...
+            GROUP BY status
+            ) AS avg_status_loan ON l.status = avg_status_loan.status
+ORDER BY a.account_id
+```
+
+</details>
+<!-- #endregion -->
+
+### Question 2 (Hard):
+Query the district_id and district_name that have the highest amount of loans for each loan status type.
```
Comment (TegveerG) [inline course-material/advanced-querying-techniques/subqueries.md:223]: variable names quote/highlight

File: course-material/advanced-querying-techniques/subqueries.md [hunk 6]
```diff
@@ -1,14 +1,316 @@
...
+    )
+
+```
+</details>
+<!-- #endregion -->
+
+
+## Wrapping Up
+
+In this section, we introduced subqueries and how they can be implemented with common clauses. This powerful, easy to understand technique spans from being a potential replacement for joins to allowing users to use results from another query as an inner query for an outer query.
```
Comment (TegveerG) [inline course-material/advanced-querying-techniques/subqueries.md:314]: The second sentence seems a bit tedious to grasp maybe because of the number of prepositions. I like bullet points in this section for brevity, if not those then rewording the second sentence would be sweet

File: course-material/advanced-querying-techniques/subqueries.md (hunk 1 above)
Comment (lfunderburk) [reply course-material/advanced-querying-techniques/subqueries.md:114]: the `` notation (`variable`) is good

File: course-material/advanced-querying-techniques/subqueries.md (hunk 2 above)
Comment (lfunderburk) [reply course-material/advanced-querying-techniques/subqueries.md:143]: `account_id` looks good

File: course-material/advanced-querying-techniques/subqueries.md (hunk 6 above)
Comment (jpjon) [reply course-material/advanced-querying-techniques/subqueries.md]: The second sentence has been revised. Please take a look and tell me what you think.

File: course-material/advanced-querying-techniques/subqueries.md [hunk 7]
```diff
@@ -1,14 +1,316 @@
...
+    jupytext_version: 1.14.6
+kernelspec:
+  display_name: jupyblog
+  language: python
+  name: python3
 ---
 
 # Using subqueries
 
+Subqueries, also known as nested queries or inner query, is a technique embedded within another query. It enables users to run a query that has some kind of relation to another query in the same SQL statement. In simpler terms, a subquery is a query within a query.
```
Comment (lfunderburk) [inline course-material/advanced-querying-techniques/subqueries.md:16]: query -> queries

or 

queries -> query

Comment (lfunderburk) [reply course-material/advanced-querying-techniques/subqueries.md:16]: Grammar feels a bit off

Subqueries (are a technique)

File: course-material/advanced-querying-techniques/subqueries.md [hunk 8]
```diff
@@ -1,14 +1,316 @@
...
+SELECT a.account_id,
+      (SELECT AVG(l.amount) 
+      FROM s1.loan AS l 
+      WHERE l.account_id = a.account_id
+      ) AS average_loan_amount
+FROM s1.account AS a
+WHERE average_loan_amount IS NOT NULL
+ORDER BY average_loan_amount
+```
+
```
Comment (lfunderburk) [inline course-material/advanced-querying-techniques/subqueries.md:142]: After you introduce the 2 queries, now show how you can easily handle large queries with JupySQL with the --save option

https://jupysql.ploomber.io/en/latest/compose.html

File: course-material/advanced-querying-techniques/subqueries.md [hunk 9]
```diff
@@ -1,14 +1,316 @@
...
+            l.amount, 
+            a.district_id 
+    FROM s1.loan AS l 
+    INNER JOIN s1.account AS a 
+      ON l.account_id = a.account_id
+    ) AS inner_query
+INNER JOIN s1.district AS d
+  ON inner_query.district_id = d.district_id
+GROUP BY d.district_name
+
```
Comment (lfunderburk) [inline course-material/advanced-querying-techniques/subqueries.md:166]: Similar comment as above, demonstrate how you can make this more reusable and clean by using the --save option

File: course-material/advanced-querying-techniques/subqueries.md [hunk 10]
```diff
@@ -1,14 +1,316 @@
...
+INNER JOIN 
+    s1.loan AS l ON a.account_id = l.account_id
+INNER JOIN (SELECT status, 
+                   AVG(amount) AS amount
+            FROM s1.loan
+            GROUP BY status
+            ) AS avg_status_loan ON l.status = avg_status_loan.status
+ORDER BY a.account_id
+```
+
```
Comment (lfunderburk) [inline course-material/advanced-querying-techniques/subqueries.md:218]: Add exercise that allows the user to practice with the --save option

https://jupysql.ploomber.io/en/latest/compose.html

File: course-material/advanced-querying-techniques/subqueries.md [hunk 11]
```diff
@@ -1,14 +1,316 @@
...
+</details>
+<!-- #endregion -->
+
+### Question 2 (Hard):
+Query the `district_id` and `district_name` that have the highest amount of loans for each loan status type.
+
+<b>Hint</b> `UNION ALL` is a clause that concatenates rows on top of each other. Use this clause to achieve the final result.
```
Comment (lfunderburk) [inline course-material/advanced-querying-techniques/subqueries.md:225]: Was UNION ALL covered previously?

Otherwise it might be too hard of a question

Can you do a brief introduction to UNION ALL at the beginning of this question, and combine it with the --save options to combine complex queries

**Review by lfunderburk — CHANGES_REQUESTED** [review]
Added observations on improving usability by allowing the user to explore existing features of JupySQL to organize queries

Incorporate

https://jupysql.ploomber.io/en/latest/compose.html

https://jupysql.ploomber.io/en/latest/api/magic-snippets.html

https://jupysql.ploomber.io/en/latest/user-guide/table_explorer.html

**Comment by jpjon** [comment]
I have adjusted the section to the above suggestions. @lfunderburk please review

File: course-material/advanced-querying-techniques/subqueries.md [hunk 12]
```diff
@@ -1,14 +1,497 @@
...
+A subquery is a technique that essentially embeds a query within another query. It enables users to run a query that has some kind of relation to another query in the same SQL statement. 
 
+Let's demonstrate how powerful subqueries are by continuing with our banking data.
 
+<!-- region -->
+
+## Install - execute this once.
+
+```{important}
+<b>Note:</b> If you are following these lessons locally and <b>not</b> on Google Colab, then there is no need to reinstall these packages.
```
Comment (lfunderburk) [inline course-material/advanced-querying-techniques/subqueries.md:25]: Change the note to say:

"The `--save` and `%sqlcmd` features used required the latest JupySQL version. Ensure you run the code below"

**Review by lfunderburk — CHANGES_REQUESTED** [review]
One final detail to change. Please address comment. 
Once this is complete I will merge.

Awesome work!

File: course-material/advanced-querying-techniques/subqueries.md (hunk 12 above)
Comment (jpjon) [reply course-material/advanced-querying-techniques/subqueries.md]: The note has been udpated.

**Review by lfunderburk — APPROVED** [review]
Merging to main. Awesome work!
